File: driveway_check_rhino/src/Part1_Identify/Core/SkeletonGraph.cs
using System.Collections.Generic;
using System.Linq;
using NetTopologySuite.Geometries;

namespace DrivewayChecker.Identify.Core
{
    // Approximate medial axis, as a plain graph of NTS coordinates. No RhinoCommon
    // dependency -- ported from reference/core/skeleton.py's SkeletonGraph dataclass.
    public sealed class SkeletonGraph
    {
        public IReadOnlyList<Coordinate> Nodes { get; }
        public IReadOnlyList<(int A, int B)> Edges { get; }

        public SkeletonGraph(IReadOnlyList<Coordinate> nodes, IReadOnlyList<(int A, int B)> edges)
        {
            Nodes = nodes;
            Edges = edges;
        }

        public int[] ComputeDegrees()
        {
            var degrees = new int[Nodes.Count];
            foreach (var (a, b) in Edges)
            {
                degrees[a]++;
                degrees[b]++;
            }
            return degrees;
        }

        public int Degree(int nodeIndex)
        {
            int count = 0;
            foreach (var (a, b) in Edges)
                if (a == nodeIndex || b == nodeIndex) count++;
            return count;
        }

        // Nodes where the skeleton splits (degree >= 3) -- candidate boundaries
        // between functional zones (e.g. aisle meets stall).
        public List<int> BranchNodes()
        {
            var degrees = ComputeDegrees();
            return Enumerable.Range(0, Nodes.Count).Where(i => degrees[i] >= 3).ToList();
        }

        // Degree-1 nodes -- candidate drive-strip terminations (street connection)
        // or dead-end turnarounds.
        public List<int> EndpointNodes()
        {
            var degrees = ComputeDegrees();
            return Enumerable.Range(0, Nodes.Count).Where(i => degrees[i] == 1).ToList();
        }

        public Dictionary<int, List<int>> BuildAdjacency()
        {
            var adjacency = new Dictionary<int, List<int>>();
            for (int i = 0; i < Nodes.Count; i++) adjacency[i] = new List<int>();
            foreach (var (a, b) in Edges)
            {
                adjacency[a].Add(b);
                adjacency[b].Add(a);
            }
            return adjacency;
        }
    }
}

File: driveway_check_web/src/lib/identify/skeleton-graph.ts
/** Approximate medial axis, as a plain graph of points.
 *
 * Ported from Part1_Identify/Core/SkeletonGraph.cs. No geometry library here —
 * this is a data structure plus counting. */

export type Pt = { x: number; y: number };

export type Edge = [number, number];

export class SkeletonGraph
{
	constructor(
		readonly nodes: readonly Pt[],
		readonly edges: readonly Edge[]
	) {}

	compute_degrees(): number[]
	{
		const degrees = new Array<number>(this.nodes.length).fill(0);
		for (const [a, b] of this.edges)
		{
			degrees[a]++;
			degrees[b]++;
		}
		return degrees;
	}

	/** Nodes where the skeleton splits — candidate boundaries between zones. */
	branch_nodes(): number[]
	{
		const degrees = this.compute_degrees();
		return degrees.map((_, i) => i).filter((i) => degrees[i] >= 3);
	}

	/** Degree-1 nodes — drive terminations or dead-end turnarounds. */
	endpoint_nodes(): number[]
	{
		const degrees = this.compute_degrees();
		return degrees.map((_, i) => i).filter((i) => degrees[i] === 1);
	}

	/** Every index gets an entry, including isolated nodes. A lazily-filled map
	 * would hand callers undefined where the C# returns an empty list. */
	build_adjacency(): Map<number, number[]>
	{
		const adjacency = new Map<number, number[]>();
		for (let i = 0; i < this.nodes.length; i++) adjacency.set(i, []);
		for (const [a, b] of this.edges)
		{
			adjacency.get(a)!.push(b);
			adjacency.get(b)!.push(a);
		}
		return adjacency;
	}
}

export function distance(a: Pt, b: Pt): number
{
	return Math.hypot(a.x - b.x, a.y - b.y);
}

/** Summed chord length along a node path. */
export function path_length(graph: SkeletonGraph, path: readonly number[]): number
{
	let total = 0;
	for (let i = 0; i + 1 < path.length; i++)
	{
		total += distance(graph.nodes[path[i]], graph.nodes[path[i + 1]]);
	}
	return total;
}
